index.js

```diff
@@ -85,8 +81,213 @@ async function SDK ({
+    return corestore.inner._deriveSecret(namespace, name)
+  }
+
+  function close (cb) {
+    swarm.close().then(cb, cb)
+  }
+
+  function makeHyperdrive (key, opts) {
+    const { persist } = opts
+
+    let driveStorage = corestore
+    if (!persist) {
+      driveStorage = RAM
+    } else if (opts.storage) {
+      driveStorage = opts.storage(key)
+    } else {
+      driveStorage = corestore
+    }
+
+    const drive = hyperdrive(driveStorage, key, opts)
+
+    if (driveStorage !== corestore) {
+      drive.ready(() => {
+        for (const core of drive.corestore.store.list().values()) {
+          trackMemoryCore(core)
+        }
+        drive.corestore.store.on('feed', (core) => {
+          trackMemoryCore(core)
+        })
+      })
+    }
+
+    return drive
+  }
+
+  function makeHypercore (name, key, opts) {
+    const { persist } = opts
+    let coreStorage
+    if (!persist) {
+      coreStorage = RAM
+    } else if (opts.storage) {
+      coreStorage = opts.storage(key)
+    }
+
+    let core = null
+
+    // If sotrage was passed in the opts, use it. Else use the corestore
+    if (coreStorage) {
+      // We only want to generate keys if we have a custom storage
+      // Else the corestore does fancy key storage for us
+      if (!key) {
+        const { publicKey, secretKey } = crypto.keyPair()
+        key = publicKey
+        opts.secretKey = secretKey
+      }
+      core = hypercore(coreStorage, key, opts)
+
+      trackMemoryCore(core)
+    } else {
+      if (key) {
+        // If a dat key was provided, get it from the corestore
+        core = corestore.get({ ...opts, key })
+      } else {
+        // If no dat key was provided, but a name was given, use it as a namespace
+        core = corestore.namespace(name).default(opts)
+      }
+    }
+    return core
+  }
+
+  function trackMemoryCore (core) {
+    core.ready(() => {
+      cacheCore(core)
+      corestore.inner._injectIntoReplicationStreams(core)
+      corestore.inner.emit('feed', core)
+    })
+
+    core.once('close', () => {
+      uncacheCore(core)
+    })
+  }
+
+  function cacheCore (core) {
+    corestore.inner.cache.set(core.discoveryKey.toString('hex'), core)
+  }
+
+  function uncacheCore (core) {
+    corestore.inner.cache.delete(core.discoveryKey.toString('hex'))
+  }
+}
+
+async function hyperspaceBackend (opts) {
+  let {
+    corestore,
+    hyperspaceClient,
+    // persist,
+    clientOpts
+  } = opts
+
+  if (!corestore) {
+    if (!hyperspaceClient) {
+      hyperspaceClient = new HyperspaceClient(clientOpts)
+    }
+    await hyperspaceClient.ready()
+    corestore = hyperspaceClient.corestore()
+  }
+
+  await hyperspaceClient.network.ready()
+  const swarm = hyperspaceClient.network
+  const keyPair = hyperspaceClient.network.keyPair
+
+  return {
+    corestore,
+    swarm,
+    keyPair,
+    deriveSecret,
+    close,
+    makeHyperdrive,
+    makeHypercore
+  }
+
+  async function deriveSecret (namespace, name) {
```

Comment:
This is a pretty important feature to me. 😅 Any chance we could get it in upstream? Might be good to add a note to the README about this change.